dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/AI/Prompt/PromptManager.cs

```diff
@@ -0,0 +1,203 @@
+﻿using Microsoft.Bot.Builder.M365.Exceptions;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.Orchestration;
+using Microsoft.SemanticKernel.TemplateEngine;
+
+namespace Microsoft.Bot.Builder.M365.AI.Prompt
+{
+    public class PromptManager<TState> : IPromptManager<TState> where TState : TurnState
+    {
+        private string? _promptsFolder;
+        private readonly Dictionary<string, PromptTemplate> _templates;
+        private readonly Dictionary<string, TemplateFunctionEntry<TState>> _functions;
+
+        public PromptManager(string? promptsFolder = null)
+        {
+            if (promptsFolder != null)
+            {
+                _VerifyDirectoryExists(promptsFolder);
+                _promptsFolder = promptsFolder;
+            }
+
+            _templates = new Dictionary<string, PromptTemplate>();
+            _functions = new Dictionary<string, TemplateFunctionEntry<TState>>();
+        }
+
+        /// <inheritdoc />
+        public IPromptManager<TState> AddFunction(string name, PromptFunction<TState> promptFunction, bool allowOverrides = false)
+        {
+            if (!_functions.ContainsKey(name) || allowOverrides)
+            {
+                _functions[name] = new TemplateFunctionEntry<TState>(promptFunction, allowOverrides);
+            } else
+            {
+                if (_functions.TryGetValue(name, out var entry))
+                {
+                    if (entry.AllowOverrides)
+                    {
+                        entry.Handler = promptFunction;
+                    } else
+                    {
+                        throw new PromptManagerException($"Attempting to update a previously registered function `{name}`");
```

Comment:
there isn't a way to unregister a function, but there's a way to override a function. If the user requires, they could override a function to return the empty string. When the prompt is rendered it would be as if the function wasn't part of the prompt template in the first place.  

This is a good point to think about, but to maintain parity with JS SDK I won't be including a method to unregister the function.